File: internal/routes/tracking_operations_test.go
package routes

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	modelsv2 "github.com/ClashKingInc/ClashKingAPI/internal/models/v2"
	apptypes "github.com/ClashKingInc/ClashKingAPI/internal/utils"
	"github.com/gofiber/fiber/v2"
)

func TestTrackingWindowsAreFixedAndBounded(t *testing.T) {
	tests := map[string]struct {
		duration time.Duration
		bucket   time.Duration
		points   int
	}{
		"15m": {15 * time.Minute, 15 * time.Second, 60},
		"1h":  {time.Hour, time.Minute, 60},
		"6h":  {6 * time.Hour, 5 * time.Minute, 72},
		"24h": {24 * time.Hour, 15 * time.Minute, 96},
	}
	for value, expected := range tests {
		window, err := parseTrackingWindow(value)
		if err != nil {
			t.Fatalf("parseTrackingWindow(%q): %v", value, err)
		}
		if window.duration != expected.duration || window.bucket != expected.bucket {
			t.Errorf("parseTrackingWindow(%q) = %+v, want duration %s and bucket %s", value, window, expected.duration, expected.bucket)
		}
		if points := int(window.duration / window.bucket); points != expected.points {
			t.Errorf("parseTrackingWindow(%q) has %d points, want %d", value, points, expected.points)
		}
	}

	if _, err := parseTrackingWindow("2d"); err == nil {
		t.Fatal("expected arbitrary tracking window to be rejected")
	}
	window, err := parseTrackingWindow("")
	if err != nil || window.name != "1h" {
		t.Fatalf("empty window = %+v, %v; want 1h default", window, err)
	}
}

func TestTrackingSeriesFiltersAreExactNames(t *testing.T) {
	for _, value := range []string{"globalclans", "globalclans.priority", "war-discovery", "tracked_clans"} {
		if got, err := parseTrackingSeriesFilter("script", value); err != nil || got != value {
			t.Errorf("filter %q = %q, %v", value, got, err)
		}
	}
	for _, value := range []string{"globalclans; DROP TABLE tracking_domain_stats", "*", strings.Repeat("a", 101)} {
		if _, err := parseTrackingSeriesFilter("domain", value); err == nil {
			t.Errorf("expected unsupported filter %q to fail", value)
		}
	}
}

func TestTrackingTargetDeltaHandlesLoopRolloverAndRestart(t *testing.T) {
	previous := &trackingTargetSnapshot{runID: 4, count: 100, cycle: 7, processed: 90}
	tests := []struct {
		name     string
		current  trackingTargetSnapshot
		previous *trackingTargetSnapshot
		want     int64
	}{
		{name: "same cycle", current: trackingTargetSnapshot{runID: 4, count: 100, cycle: 7, processed: 97}, previous: previous, want: 7},
		{name: "one rollover", current: trackingTargetSnapshot{runID: 4, count: 100, cycle: 8, processed: 6}, previous: previous, want: 16},
		{name: "multiple rollovers", current: trackingTargetSnapshot{runID: 4, count: 100, cycle: 10, processed: 6}, previous: previous, want: 216},
		{name: "process restart", current: trackingTargetSnapshot{runID: 5, count: 100, cycle: 1, processed: 8}, previous: previous, want: 8},
		{name: "first observation", current: trackingTargetSnapshot{runID: 5, count: 100, cycle: 1, processed: 8}, want: 8},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			if got := trackingTargetDelta(test.current, test.previous); got != test.want {
				t.Fatalf("trackingTargetDelta() = %d, want %d", got, test.want)
			}
		})
	}
}

func TestTrackingDomainCalculationsStayInAPI(t *testing.T) {
	now := time.Date(2026, 8, 27, 12, 0, 0, 0, time.UTC)
	previousRunID := int64(1)
	previousCount := 100
	previousCycle := int64(2)
	previousProcessed := 40
	latestError := "proxy timeout"
	latestErrorAt := now.Add(-3 * time.Second)
	raw := trackingRawDomain{
		intervalStart: now.Add(-10 * time.Second), intervalEnd: now,
		runID: 1, script: "globalclans", name: "globalclans.priority",
		lastError: &latestError, lastReadyChange: &latestErrorAt, requests: 50, writes: 20, errors: 5,
		requestLatencyMS: 1_000, queueDepth: 7, reportedHealthy: true,
		processingCount: 10, totalProcessTimeMS: 500,
		storeBatches: 4, storeRowsRequested: 30, storeRowsAffected: 28, storeDurationMS: 200,
		targetCount: 100, targetCycle: 2, targetProcessed: 50,
		previousRunID: &previousRunID, previousTargetCount: &previousCount,
		previousTargetCycle: &previousCycle, previousTargetProcessed: &previousProcessed,
	}
	got := buildTrackingDomainState(raw, now)
	if got.RequestsPerSecond != 5 || got.ErrorRate != 0.1 || got.AverageRequestLatencyMS != 20 {
		t.Fatalf("request metrics = rps %v, error rate %v, latency %v", got.RequestsPerSecond, got.ErrorRate, got.AverageRequestLatencyMS)
	}
	if got.WritesPerSecond != 2 || got.Database.AverageStoreDurationMS != 50 {
		t.Fatalf("write/store metrics = writes/s %v, store duration %v", got.WritesPerSecond, got.Database.AverageStoreDurationMS)
	}
	if got.Targets == nil || got.Targets.TargetsPerSecond != 1 || got.Targets.CompletionPercentage != 50 {
		t.Fatalf("target metrics = %+v", got.Targets)
	}
	if got.Targets.EstimatedSecondsRemaining == nil || *got.Targets.EstimatedSecondsRemaining != 50 {
		t.Fatalf("target ETA = %+v", got.Targets)
	}
	if got.LatestError == nil || got.LatestError.Message != latestError || !got.LatestError.Timestamp.Equal(latestErrorAt) {
		t.Fatalf("latest error = %+v", got.LatestError)
	}
}

func TestTrackingQueueOnlyDomainHasMetricsWithoutTargetProgress(t *testing.T) {
	now := time.Date(2026, 8, 27, 12, 0, 0, 0, time.UTC)
	latestError := "queue worker timeout"
	latestErrorAt := now.Add(-5 * time.Second)
	raw := trackingRawDomain{
		intervalStart: now.Add(-20 * time.Second), intervalEnd: now,
		runID: 3, script: "reminders", name: "reminders.mobile-reconciliation",
		requests: 40, writes: 10, errors: 2, requestLatencyMS: 800,
		queueDepth: 12, reportedHealthy: false, processingCount: 5,
		totalProcessTimeMS: 250, targetCount: 0, lastError: &latestError,
		lastReadyChange: &latestErrorAt,
	}
	got := buildTrackingDomainState(raw, now)
	if got.Domain != raw.name || got.Targets != nil {
		t.Fatalf("queue domain = %+v; want arbitrary name and null target progress", got)
	}
	if got.QueueDepth != 12 || got.ProcessingCount != 5 || got.AverageProcessingDurationMS != 50 || got.RequestsPerSecond != 2 || got.WritesPerSecond != .5 || got.ErrorCount != 2 || got.AverageRequestLatencyMS != 20 {
		t.Fatalf("queue metrics = %+v", got)
	}
	if got.Health.Stale || got.Health.Healthy || got.Health.AgeSeconds != 0 || got.LatestError == nil || !got.LatestError.Timestamp.Equal(latestErrorAt) {
		t.Fatalf("queue freshness/error = health %+v, error %+v", got.Health, got.LatestError)
	}
	payload, err := json.Marshal(got)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(payload), `"targets":null`) {
		t.Fatalf("queue domain JSON = %s; want explicit null target progress", payload)
	}
}

func TestTrackingArbitraryDomainNameIsPreserved(t *testing.T) {
	now := time.Date(2026, 8, 27, 12, 0, 0, 0, time.UTC)
	for _, domain := range []string{
		"globalclans.priority", "globalclans.non_priority",
		"war-discovery.active", "war-discovery.dormant", "war-discovery.finalization",
		"cwl.groups", "war-archiver",
		"reminders.events", "reminders.war-jobs", "reminders.mobile-reconciliation", "reminders.raid", "reminders.fixed",
		"mobilepush.events", "trackedplayers", "future-domain.with_any-name",
	} {
		raw := trackingRawDomain{intervalStart: now.Add(-time.Second), intervalEnd: now, script: "worker", name: domain, reportedHealthy: true}
		if got := buildTrackingDomainState(raw, now); got.Domain != domain {
			t.Errorf("domain %q became %q", domain, got.Domain)
		}
	}
}

func TestTrackingDomainFailureMakesFreshProcessUnhealthy(t *testing.T) {
	now := time.Date(2026, 8, 27, 12, 0, 0, 0, time.UTC)
	processes := []modelsv2.TrackingProcessState{{
		Script: "reminders", Health: trackingHealth(true, now, now),
	}}
	domains := []modelsv2.TrackingDomainState{
		{Script: "reminders", Domain: "reminders.events", Health: trackingHealth(true, now, now)},
		{Script: "reminders", Domain: "reminders.raid", Health: trackingHealth(false, now, now)},
	}
	rollUpTrackingProcessHealth(processes, domains)
	if processes[0].Health.Healthy || processes[0].Health.Stale || processes[0].Health.ReportedHealthy {
		t.Fatalf("fresh process with failed domain should be unhealthy: %+v", processes[0].Health)
	}
}

func TestTrackingHealthMarksOldObservationsStale(t *testing.T) {
	now := time.Date(2026, 8, 27, 12, 0, 0, 0, time.UTC)
	fresh := trackingHealth(true, now.Add(-trackingStaleAfter), now)
	if fresh.Stale || !fresh.Healthy {
		t.Fatalf("observation at threshold should be fresh: %+v", fresh)
	}
	stale := trackingHealth(true, now.Add(-trackingStaleAfter-time.Millisecond), now)
	if !stale.Stale || stale.Healthy || !stale.ReportedHealthy {
		t.Fatalf("old observation should be stale but retain reported health: %+v", stale)
	}
	unhealthy := trackingHealth(false, now, now)
	if unhealthy.Healthy || unhealthy.Stale {
		t.Fatalf("fresh reported failure should be unhealthy, not stale: %+v", unhealthy)
	}
}

func TestTrackingAdminEndpointsRequireServiceToken(t *testing.T) {
	deps := apptypes.Deps{Config: apptypes.Config{APIBotToken: "admin-secret"}}
	app := fiber.New(fiber.Config{ErrorHandler: apptypes.ErrorHandler})
	app.Get("/v2/admin/tracking/summary", authBot(deps, trackingOperationsSummary(deps)))

	for _, authorization := range []string{"", "Bearer wrong-secret"} {
		req := httptest.NewRequest(http.MethodGet, "/v2/admin/tracking/summary", nil)
		if authorization != "" {
			req.Header.Set(fiber.HeaderAuthorization, authorization)
		}
		response, err := app.Test(req)
		if err != nil {
			t.Fatal(err)
		}
		if response.StatusCode != fiber.StatusUnauthorized {
			t.Fatalf("authorization %q returned %d, want 401", authorization, response.StatusCode)
		}
	}

	req := httptest.NewRequest(http.MethodGet, "/v2/admin/tracking/summary", nil)
	req.Header.Set(fiber.HeaderAuthorization, "Bearer admin-secret")
	response, err := app.Test(req)
	if err != nil {
		t.Fatal(err)
	}
	if response.StatusCode != fiber.StatusServiceUnavailable {
		t.Fatalf("valid admin token returned %d, want handler's 503 without a test store", response.StatusCode)
	}
	if got := response.Header.Get(fiber.HeaderCacheControl); got != "no-store" {
		t.Fatalf("Cache-Control = %q, want no-store", got)
	}
}

func TestTrackingAdminEndpointsAreRegistered(t *testing.T) {
	app := newRegisteredRoutesTestApp()
	Register(app, apptypes.Deps{}, func(next fiber.Handler) fiber.Handler { return next })
	for _, path := range []string{"/v2/admin/tracking/summary", "/v2/admin/tracking/timeseries"} {
		if registeredRouteIndex(app, fiber.MethodGet, path) < 0 {
			t.Fatalf("expected GET %s to be registered", path)
		}
	}
}

func TestTrackingQueriesAreTimeBoundedAndParameterized(t *testing.T) {
	for name, query := range map[string]string{
		"process": trackingProcessSeriesQuery,
		"domain":  trackingDomainSeriesQuery,
	} {
		for _, fragment := range []string{"interval_end >= $1", "interval_end < $2", "LIMIT 20000"} {
			if !strings.Contains(query, fragment) {
				t.Errorf("%s series query is missing %q", name, fragment)
			}
		}
		if strings.Contains(query, "%s") || strings.Contains(query, "+ script") {
			t.Errorf("%s series query appears to interpolate caller input: %s", name, query)
		}
	}
}

File: workers/api/src/tracking-operations.test.ts
import { Effect, Schema } from "effect"
import { SqlClient } from "effect/unstable/sql"
import { describe, expect, it, vi } from "vitest"
import { TrackingSummaryResponse, TrackingTimeSeriesResponse } from "../../../packages/api-contracts/src/admin.js"
import { executeTrackingRead, parseTrackingQuery, trackingReadInternals as internals } from "./tracking-operations.js"

const now = Date.parse("2026-09-04T12:00:00Z")
const raw = {
  interval_start: new Date(now - 10_000), interval_end: new Date(now), run_id: "1", script: "globalclans", name: "globalclans.priority",
  last_success: null, last_error: "proxy timeout", last_ready_change: new Date(now - 3000), requests: "50", writes: "20", errors: "5",
  request_latency_ms: 1000, queue_depth: 7, healthy: true, processing_count: "10", total_process_time_ms: 500,
  store_batches: "4", store_rows_requested: "30", store_rows_affected: "28", store_duration_ms: 200,
  target_count: 100, target_cycle: "2", target_processed: 50, previous_run_id: "1", previous_target_count: 100, previous_target_cycle: "2", previous_target_processed: 40,
}
const noSql = {} as SqlClient.SqlClient

describe("original API tracking calculations and contracts", () => {
  it.each([["15m", 900, 15, 60], ["1h", 3600, 60, 60], ["6h", 21600, 300, 72], ["24h", 86400, 900, 96]] as const)("retains the bounded %s window", (name, duration, bucket, points) => {
    expect(internals.windows[name]).toEqual([duration, bucket])
    expect(duration / bucket).toBe(points)
    expect(parseTrackingQuery({ window: name })).toEqual({ window: name, script: "", domain: "" })
  })
  it("retains empty/default, trim, case normalization and exact script/domain filters", () => {
    expect(parseTrackingQuery({})).toEqual({ window: "1h", script: "", domain: "" })
    expect(parseTrackingQuery({ window: " 6H ", script: " war-discovery ", domain: " cwl.groups " })).toEqual({ window: "6h", script: "war-discovery", domain: "cwl.groups" })
  })
  it.each([{ window: "2d" }, { window: "__proto__" }, { script: "*" }, { domain: "x; DROP TABLE tracking_domain_stats" }, { domain: "x".repeat(101) }])("rejects invalid queries before SQL %#", async query => {
    await expect(Effect.runPromise(executeTrackingRead("timeseries", query).pipe(Effect.provideService(SqlClient.SqlClient, noSql))))
      .rejects.toMatchObject({ _tag: "InvalidRequest" })
  })
  it("preserves original rates, ETA and structured latest error", () => {
    const domain = internals.domainState(raw, now)
    expect(domain).toMatchObject({ requests_per_second: 5, error_rate: 0.1, average_request_latency_ms: 20, writes_per_second: 2,
      database: { average_store_duration_ms: 50 }, targets: { targets_per_second: 1, completion_percentage: 50, estimated_seconds_remaining: 50 },
      latest_error: { message: "proxy timeout", timestamp: "2026-09-04T11:59:57.000Z" } })
    const response = { generated_at: new Date(now).toISOString(), stale_after_seconds: 120, processes: [], domains: [domain] }
    expect(Schema.decodeUnknownSync(TrackingSummaryResponse)(response)).toEqual(response)
  })
  it("accepts queue-only schema003 null targets and keeps null error metadata", () => {
    const domain = internals.domainState({ ...raw, target_count: null, target_cycle: null, target_processed: null, last_error: null, healthy: false }, now)
    expect(domain.targets).toBeNull()
    expect(domain.latest_error).toBeNull()
    expect(domain.health).toMatchObject({ healthy: false, stale: false })
    expect(() => Schema.decodeUnknownSync(TrackingSummaryResponse)({ generated_at: new Date(now).toISOString(), stale_after_seconds: 120, processes: [], domains: [domain] })).not.toThrow()
  })
  it.each([[7, 97, "4", 7], [8, 6, "4", 16], [10, 6, "4", 216], [1, 8, "5", 8]] as const)("preserves cycle/restart delta %#", (cycle, processed, run, delta) => {
    expect(internals.targetDelta({ count: 100, cycle, processed, run, observedAt: now }, { count: 100, cycle: 7, processed: 90, run: "4", observedAt: now - 1000 })).toBe(delta)
  })
  it("retains the exact stale threshold and never estimates stale target completion", () => {
    expect(internals.health(true, new Date(now - 120_000), now)).toMatchObject({ healthy: true, stale: false })
    expect(internals.health(true, new Date(now - 120_001), now)).toMatchObject({ healthy: false, stale: true, reported_healthy: true })
    expect(internals.domainState(raw, now + 120_001).targets).not.toHaveProperty("estimated_loop_completion")
  })
  it("uses the preceding chart bucket only for progress delta, not returned points", () => {
    const point = { ...raw, bucket: new Date(now), observed_at: new Date(now + 10_000), duration_seconds: 10 }
    const previous = { ...point, bucket: new Date(now - 60_000), observed_at: new Date(now - 10_000), target_processed: 40 }
    const domains = internals.groupedDomains([previous, point], now)
    expect(domains[0]!.points).toHaveLength(1)
    expect(domains[0]!.points[0]!.targets).toMatchObject({ targets_per_second: 0.5 })
    expect(() => Schema.decodeUnknownSync(TrackingTimeSeriesResponse)({ generated_at: new Date(now).toISOString(), window: "1h", start: new Date(now).toISOString(), end: new Date(now + 3600_000).toISOString(), bucket_seconds: 60, max_points_per_series: 60, processes: [], domains })).not.toThrow()
  })
  it("keeps both queries bounded and parameterized", () => {
    for (const query of [internals.processSeries, internals.domainSeries]) {
      expect(query).toContain("interval_end >= $1 AND interval_end < $2")
      expect(query).toContain("LIMIT 20000")
    }
    expect(internals.domainSeries).toContain("($5='' OR name=$5)")
  })
  it("does not ask a provider for either tracking read", async () => {
    const outbound = vi.spyOn(globalThis, "fetch").mockRejectedValue(new Error("Unexpected provider"))
    try {
      for (const endpoint of ["summary", "timeseries"] as const) {
        const sql = { unsafe: () => Effect.succeed([]) } as unknown as SqlClient.SqlClient
        const result = await Effect.runPromise(executeTrackingRead(endpoint, {}).pipe(Effect.provideService(SqlClient.SqlClient, sql)))
        expect(result).toMatchObject({ processes: [], domains: [] })
      }
      expect(outbound).not.toHaveBeenCalled()
    } finally { outbound.mockRestore() }
  })
})
